README.md
# RelevantVids

<img src="https://github.com/andrewshvv/RelevantVids/blob/main/example.png" width="50%">


This Chrome extension is a lighthouse amidst the whirlpool of YouTube distractions, guiding your learning journey with precision and ease. When using YouTube for educational purposes, it's easy to get swept up in the sea of unrelated videos that constantly vie for your attention. This extension is an adept solution to this common dilemma.

Just set your desired learning theme in the extension, and watch as your YouTube page transforms from an overwhelming medley of content into a curated gallery aligned with your interests. It sifts through video titles, highlights the ones relevant to your set theme, and reduces the noise from irrelevant, off-topic content.

You have the freedom to change themes as your interests evolve, keeping your learning journey fresh and focused. Need a regular browsing break? Clear the theme to revert back to the standard YouTube view. In essence, this tool allows you to cut through the distraction, paving a more focused path for your YouTube learning experience.

### A bit more technical description
This Chrome extension works by manipulating the YouTube webpage's HTML based on the user's specified theme, using JavaScript and Chrome's APIs. It leverages advanced language understanding capabilities of OpenAI's powerful GPT model for this purpose.

When a theme is set by the user via the extension, an event is triggered that retrieves this theme from Chrome's local storage. The extension then seeks out the video grid containers on the YouTube page and starts monitoring them using a MutationObserver instance.

The MutationObserver watches for changes or additions to the video grid, such as the appearance of new videos. Whenever such a change is detected, the extension examines the newly added video elements and checks their relevancy against the user-specified theme.

For each new video, it extracts the title and sends it along with the user's specified theme as a request to OpenAI's GPT API. This remote server runs a language-based model to determine whether the video title is relevant to the theme. Note that, to save computational resources and API calls, previously checked videos and their results are stored in a local Map.

Once the response from the API is received, based on the relevancy of the video to the theme, the extension applies CSS manipulation to highlight relevant videos and dims the irrelevant ones, thereby altering their visibility in the video grid.

Additionally, the extension maintains an "active state" for each page of YouTube, which allows it to update the curated view dynamically as the user navigates different sections of YouTube. If the user changes the theme, the entire video grid is updated to reflect the new theme. Users also have the option to reset the videos and revert back to the standard YouTube view.

1. `gpt.js`: This file contains functions necessary for asynchronous interaction with the GPT (Generative Pretrained Transformer) model from OpenAI. It helps send requests and handle responses concerning the relevance of YouTube video titles with the user's set theme.

2. `utils.js`: This is a utility file containing various helper functions. Functions in this file support interaction with the YouTube page's Document Object Model (DOM), including finding video grid containers, checking if a video is compatible with a chosen theme, making a video active or inactive, and re-arranging the video grid.

3. `content.js`: This is the main content script for the Chrome extension. It manages the extension's behavior when the DOM is loaded or changed and listens for changes in user specified theme. It also handles interactions with YouTube's video grid container and individual videos, checking their relevancy to the selected theme and altering their visibility accordingly. It fetches the results from the OpenAI GPT model by using functions imported from the `gpt.js` file.


CHANGELOG.md
# CHANGELOG

```txt
Summary
  1. document grouping follow 'SemVer2.0' protocol
  2. use 'PATCH' as a minimum granularity
  3. use concise descriptions
  4. type: feat \ fix \ update \ perf \ remove \ docs \ chore
  5. version timestamp follow the yyyy.MM.dd format
```

## 0.0.0 [2023.06.25]

- feat: initial
- feat: generator by ![create-chrome-ext](https://www.npmjs.com/package/create-chrome-ext)


src/manifest.js
import {defineManifest} from '@crxjs/vite-plugin'

export default defineManifest({
  name: 'create-chrome-ext',
  description: '',
  version: '0.0.0',
  manifest_version: 3,
  icons: {
    16: 'img/logo-16.png',
    32: 'img/logo-34.png',
    48: 'img/logo-48.png',
    128: 'img/logo-128.png',
  },
  action: {
    default_popup: 'popup.html',
    default_icon: 'img/logo-48.png',
  },
  options_page: 'options.html',
  background: {
    service_worker: 'src/background/index.js',
    type: 'module',
  },
  content_scripts: [
    {
      matches: [
        'http://*.youtube.com/*',
        'https://*.youtube.com/*'
      ],
      css: ['src/content/style.css'],
      js: ['src/content/index.js'],
    },
  ],
  web_accessible_resources: [
    {
      resources: ['img/logo-16.png', 'img/logo-34.png', 'img/logo-48.png', 'img/logo-128.png', 'src/content/style.css'],
      matches: [
        'http://*.youtube.com/*',
        'https://*.youtube.com/*'
      ],
    },
  ],
  permissions: [
    "activeTab",
    "tabs",
    "storage",
    "webNavigation",
  ],
})


src/content/gpt.js
import Bottleneck from "bottleneck";
import {Configuration, OpenAIApi} from "openai";


export {enqueueRequest, totalTokens, limiter}

let totalTokens = 0;

async function checkVideo(data) {
  let title = data["title"];
  let theme = data["theme"];

  let content = await gptFilter(title, theme);
  content = JSON.parse(content.content)
  // console.log(content.response, title.slice(0,40));
  return content.response
  // return Math.random() < 0.1;
  // return false;
}

async function gptFilter(description, theme) {
  try {
    const chatCompletion = await openai.createChatCompletion({
      model: "gpt-3.5-turbo-0613",
      temperature: 0,
      messages: [
        {
          role: "system",
          content: "As an AI YouTube assistant, your task is based on the YouTube video description we provide and the category we designate. Determine if the video matches the provided category. If the description aligns, issue a response in JSON format as {\"response\": true}, and if not, respond with {\"response\": false}.\n",
        },
        {role: "user", content: `Here is your first task \n Description: ${description} \n Category: ${theme}`},
      ],

    });

    totalTokens += chatCompletion.data.usage.total_tokens;

    return chatCompletion.data.choices[0].message;

  } catch (error) {
    console.error("Error:", error);
  }
}

// Create a new limiter with a limit of 200 requests per minute
const limiter = new Bottleneck({
  minTime: (60 * 1000) / 3500, // 60,000 ms in a minute divided by 200 RPM
});

// Set the OpenAI API
let configuration = new Configuration({
  apiKey: "",
});

delete configuration.baseOptions.headers['User-Agent'];
const openai = new OpenAIApi(configuration);

// A wrapper function to handle retries and errors
async function makeRequestWithRetries(data, maxRetries = 3) {
  let retries = 0;
  while (retries < maxRetries) {
    try {
      return await checkVideo(data);
    } catch (error) {
      retries++;
      console.error(`Request failed: ${error}. Retrying... (${retries}/${maxRetries})`);
      // Exponential backoff delay before retrying the request
      await new Promise((resolve) => setTimeout(resolve, Math.pow(2, retries) * 1000));
    }
  }
  throw new Error(`Request failed after ${maxRetries} retries`);
}

// The enqueue function that schedules a request using the limiter
function enqueueRequest(data) {
  // console.log("added to queue", data);
  return limiter.schedule(() => makeRequestWithRetries(data));
}


src/content/utils.js
export {
  rearrangeVideoGrid,
  makeInactive,
  makeActive,
  isActive,
  CONTAINER_SELECTOR,
  GRID_SELECTOR,
  INFO_SELECTOR,
  isItVideoGridContainer,
  findVideoGridContainer,
  getYouTubePageName,
  checkVideoElement,
  findVideoContainerByHref,
  findVideoContainerByElement,
  checkIfVideoLinkChanged
};

// Constants
const CONTAINER_SELECTOR = "ytd-rich-item-renderer";
const GRID_SELECTOR = "contents"
const INFO_SELECTOR = "video-title-link"
const VIDEO_ROW_SELECTOR = "ytd-rich-grid-row"

// The `rearrangeVideos()` function rearranges video elements on a webpage based on their visibility.
// It first collects all video elements from the webpage and stores them in an array along with their visibility status.
// It then assigns an index value to each video object based on its visibility.
//
// After this, the function creates a sorted copy of the videos array and sorts the video elements by their visibility
// using the `sort()` function. During this sorting process, the `swap()` function is called to physically swap the
// position of the elements in the DOM (Document Object Model) based on their assigned index. The result is that
// visible videos are moved to the front and hidden videos are moved to the back, effectively rearranging the
// order of the video elements based on their visibility.
function rearrangeVideoGrid() {
  const rows = document.querySelectorAll(VIDEO_ROW_SELECTOR);

  let videos = [];
  rows.forEach(row => {
    const videoElements = Array.from(row.querySelectorAll(CONTAINER_SELECTOR));
    videoElements.forEach((video) => {
      videos.push({
        element: video,
      });
    });
  });

  let index = 0;
  videos.forEach((video) => {
    video.index = isActive(video.element) ? videos.length - index : 0;
    index++;
  });

  videos.sort((vA, vB) => {
    const a = vA.index;
    const b = vB.index;

    if (a > b) {
      swap(vA.element, vB.element);
    }

    return b - a;
  });
}

function swap(node1, node2) {
  // Create a temporary placeholder element
  const placeholder = document.createElement('div');

  // Insert the placeholder before node1 to mark its location
  node1.parentNode.insertBefore(placeholder, node1);

  // Move node1 to node2's location
  node2.parentNode.insertBefore(node1, node2);

  // Move node2 to the placeholder's location (original node1 location)
  placeholder.parentNode.insertBefore(node2, placeholder);

  // Remove the placeholder element
  placeholder.parentNode.removeChild(placeholder);
}

function isActive(element) {
  return !element.classList.contains('disable-pointer-events');
}

function makeInactive(element) {
  // Check if the element is currently inactive
  const isInactive = element.classList.contains('disable-pointer-events');
  if (isInactive) return;

  // Toggle the disable-pointer-events class
  element.classList.add('disable-pointer-events');

  // Update focus and tabindex for interactive elements
  const focusableElements = element.querySelectorAll('a[href], input, button, select, textarea');
  for (const focusable of focusableElements) {
    // Save the old tabindex and prevent focus
    const oldTabIndex = focusable.getAttribute('tabindex') || '0';
    focusable.setAttribute('data-old-tabindex', oldTabIndex);
    focusable.setAttribute('tabindex', '-1');
  }

  // element.style.display = 'none';
}

function makeActive(element) {
  // Check if the element is currently inactive
  const isInactive = element.classList.contains('disable-pointer-events');
  if (!isInactive) return;

  // Toggle the disable-pointer-events class
  element.classList.remove('disable-pointer-events');

  // Update focus and tabindex for interactive elements
  const focusableElements = element.querySelectorAll('a[href], input, button, select, textarea');
  for (const focusable of focusableElements) {
    // Restore the old tabindex
    const oldTabIndex = focusable.getAttribute('data-old-tabindex') || '0';
    focusable.setAttribute('tabindex', oldTabIndex);
    focusable.removeAttribute('data-old-tabindex');
  }

  // element.style.display = '';
}


function isItVideoGridContainer(node) {
  // Check if the current node has the id 'contents'
  if (node.id === 'contents') {
    // Check if the parent node exists and has the correct class
    let parentNodeType = node.parentNode && node.parentNode.getAttribute('class');
    if (
      parentNodeType &&
      (parentNodeType.includes('style-scope ytd-two-column-browse-results-renderer') ||
        parentNodeType.includes('style-scope ytd-watch-next-secondary-results-renderer'))
    ) {
      // Check if the child node exists and has the correct class
      if (
        node.children &&
        node.children.length > 0
      ) {
        let childNodeType = node.children[0].getAttribute('class');
        if (childNodeType &&
          (childNodeType.includes('style-scope ytd-rich-grid-renderer') ||
            childNodeType.includes('style-scope ytd-item-section-renderer'))) {
          return true;
        }
      }
    }
  }

  // If the conditions are not met, return false
  return false;
}

function findVideoGridContainer() {
  let nodes = document.querySelectorAll('#contents');
  if (!nodes) return null;

  for (let node of nodes) {
    // Check if the correct parent and child nodes exist.
    if (!isItVideoGridContainer(node)) continue;
    return node;
  }

  // If no matching node is found, return null
  return null;
}

function getYouTubePageName() {
  const url = window.location.href;
  let page = null;

  try {
    let urlObj = new URL(url);
    if (urlObj.pathname === "/") {
      page = "main";
    } else if (urlObj.pathname.startsWith("/watch")) {
      page = "watch";
    }
  } catch (_) { /* Failed to construct URL object, URL might not be valid */
  }

  return page;
}

function checkVideoElement(node) {
  // The first check is for the case when the added node itself has an ID 'video-title-link'.
  // This means the video element is directly added to the DOM and needs to be processed.
  if (node.nodeType === Node.ELEMENT_NODE) {
    if (node.id === INFO_SELECTOR) {
      // console.log("VIDEO ELEMENT APPEARED", node);
      return node
    } else {
      let element = node.querySelector(`#${INFO_SELECTOR}`);
      if (element) {
        // console.log("VIDEO ELEMENT part of", node, element);
        return element;
      }
    }
  }

  return null;
}

function checkIfVideoLinkChanged(node) {
  if (node.nodeType === Node.ELEMENT_NODE) {
    if (node.id === "video-title") {
      return node.parentNode
    }
  }

  return null;
}

function findVideoContainerByHref(href) {
  let link = findLinkElement(href);
  if (!link) return null;
  return findVideoContainerByElement(link);
}

function findVideoContainerByElement(element) {
  let currentElement = element.parentNode;

  while (currentElement) {
    if (currentElement.nodeName.toLowerCase() === CONTAINER_SELECTOR) {
      return currentElement;
    }

    currentElement = currentElement.parentNode;
  }

  return null;
}

function findLinkElement(href) {
  let allAnchorTags = Array.from(document.querySelectorAll('a'));

  let matchedElement = allAnchorTags.filter(el => {
    return el.getAttribute('href') === href;
  });

  return matchedElement.length > 0 ? matchedElement[0] : null;
}


src/content/index.js
import {enqueueRequest, totalTokens} from "./gpt";
import {
  rearrangeVideoGrid,
  makeInactive,
  makeActive,
  INFO_SELECTOR,
  isItVideoGridContainer,
  findVideoGridContainer,
  getYouTubePageName,
  isActive,
  checkVideoElement,
  findVideoContainerByHref,
  checkIfVideoLinkChanged
} from "./utils";

console.info('chrome-ext template-react-js content script');

const RUB_USD = 80;
const TOKEN_PRICE = 0.0015;

export {}

let context = {
  theme: null,
  states: new Map(),
  activeState: null,
}

function getCurrentState() {
  return context.activeState;
}

chrome.storage.local.get("theme", function (result) {
  updateTheme(result.theme);
});

chrome.storage.onChanged.addListener(function (changes, namespace) {
  for (let key in changes) {
    let change = changes[key];
    if (key === 'theme') {
      updateTheme(change.newValue);
      return;
    }
  }
});

chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
  if (request.message === 'ON_URL_CHANGED') {
    if (!context.states.has(getYouTubePageName())) return;
    context.activeState = context.states.get(getYouTubePageName());
    // console.log("NEW ACTIVE GRID", getYouTubePageName(), context.activeState.gridContainer);
    resyncVideoStates();
    // resyncVideoStates();
  }
});

function updateTheme(theme) {
  if (theme === "") {
    context.theme = null;
    resetVideos(true);
  } else {
    context.theme = theme;
    resetVideos(false);

    let state = getCurrentState();
    if (!state) return;

    for (const video of state.videos.values()) {
      checkVideo(video);
    }
  }
}

function resetVideos(active = false) {
  let state = getCurrentState();
  if (!state) return;

  for (const video of state.videos.values()) {
    video.status = "init";
    video.active = active;

    let element = findVideoContainerByHref(video.url);
    if (!element) return;

    if (active) {
      makeActive(element);
    } else {
      makeInactive(element);
    }
  }
}

function checkVideo(video) {
  if (video.status !== "init") return;
  video.status = "queue";

  const data = {
    "title": video.title,
    "theme": context.theme,
  };

  enqueueRequest(data)
    .then((response) => {
      video.status = "checked";
      video.active = response;

      syncVideoVisibility(video);
    })
    .catch((error) => {
      video.status = "init";
      console.error('Error:', error);
    });
}

// Resync the visibility of videos based on their active status
function resyncVideoStates() {
  let state = getCurrentState();
  if (!state) return;
  for (let video of state.videos.values()) {
    syncVideoVisibility(video);
  }
}

function createOrGetVideo(infoElem, state) {
  let videos = state.videos;
  let url = infoElem.getAttribute("href");
  if (videos.has(url)) {
    return [videos.get(url), false];
  }

  let video = {
    title: infoElem.getAttribute("aria-label"),
    url: url,
    status: "init",
    active: context.theme == null,
  };

  console.log("new video", video.title.slice(0, 40));
  videos.set(url, video);
  return [video, true];
}

function syncVideoVisibility(video) {
  let element = findVideoContainerByHref(video.url);
  if (!element) return;

  if (video.active && !isActive(element)) {
    makeActive(element);
    rearrangeVideoGrid();
    console.log("active", video.title.slice(0, 40));
  } else if (!video.active && isActive(element)) {
    makeInactive(element);
    rearrangeVideoGrid();
  }
}

// Perform a full update of the video grid by processing each video element
function fullUpdate(videoGrid, state) {
  const elements = videoGrid.querySelectorAll(`#${INFO_SELECTOR}`);

  elements.forEach((element) => {
    let [video, _] = createOrGetVideo(element, state);
    syncVideoVisibility(video);
  });
}

function DOMContentLoaded() {
  // Wait for the main container for video to appear in DOM.
  listenForVideoGrid();
}

function addNewState(gridContainer) {
  let currentState = {
    videos: new Map(),
    gridContainer: gridContainer,
  };
  context.states.set(getYouTubePageName(), currentState);
  context.activeState = currentState;
  // console.log("NEW ACTIVE GRID", getYouTubePageName(), context.activeState.gridContainer);
  return currentState;
}

function listenForVideoGrid() {
  const startSearchForMainVideoGrid = () => {
    const videoGrid = findVideoGridContainer();
    if (videoGrid) {
      let state = addNewState(videoGrid);
      listeningForNewVideos(videoGrid, state);
      fullUpdate(videoGrid, state);
    }

    // In case if page changes, we have to pick up new video grid.
    const domObserver = new MutationObserver((mutations, observer) => {
      for (const mutation of mutations) {
        // Detect if an element has just appeared
        if (mutation.type === 'childList') {
          for (const node of mutation.addedNodes) {
            if (!isItVideoGridContainer(node)) continue;
            let state = addNewState(node);
            listeningForNewVideos(node, state);
            fullUpdate(node, state);

            // After two containers disconnect
            if (context.states.size >= 2) observer.disconnect();
          }
        }
      }
    });

    domObserver.observe(document.body, {childList: true, subtree: true});
  };

  // Start waiting for the videoGrid to appear in the DOM
  startSearchForMainVideoGrid();
}

// Start observing the changes in the current active video grid container
// i.e. listening for new videos to appear on page.
function listeningForNewVideos(videoGridContainer, state) {
  const observerOptions = {
    childList: true,
    attributes: false,
    subtree: true,
  };

  const observer = new MutationObserver((mutations) => {
    for (const mutation of mutations) {
      if (mutation.type !== 'childList') return;

      // For the case when we return back, and youtube changes
      // the content of video container
      let link = checkIfVideoLinkChanged(mutation.target);
      if (link) {
        let url = link.getAttribute("href");
        let video = state.videos.get(url);
        if (video) syncVideoVisibility(video);
      }

      let element = checkVideoElement(mutation.target);
      if (element !== null) {
        // console.log("ELEMENT", element);
        let [video, justCreated] = createOrGetVideo(element, state);
        syncVideoVisibility(video);

        if (justCreated) {
          checkVideo(video);
        }
      }
    }
  });

  // Start observing the videoGridContainer with the specified configuration
  observer.observe(videoGridContainer, observerOptions);
  return observer;
}

// State info print
setInterval(function () {
  let queue = 0;
  let init = 0;
  let checked = 0;
  let active = 0;
  let inactive = 0;

  // resyncVideoStates();
  for (const video of getCurrentState().videos.values()) {
    if (video.active) {
      // console.log(video.title, findVideoContainerByHref(video.url));
      active++;
    }
    if (!video.active) inactive++;
    if (video.status === "init") init++;
    if (video.status === "queue") queue++;
    if (video.status === "checked") checked++;

  }

  console.log(`state ${(RUB_USD * TOKEN_PRICE * totalTokens / 1000).toFixed(2)},${totalTokens} ${init}/${queue}/${checked} ${active}/${inactive}/${active + inactive}`, getCurrentState());
}, 10000);


if (document.readyState === 'loading') {
  document.addEventListener('DOMContentLoaded', DOMContentLoaded);
} else {
  DOMContentLoaded();
}
